Add fields to the correct location, allow creating new block

kienstra opened the pull request:
Updated December 1st

#### Changes
* Can create a new block
* Can add a field to the 'Editor' or 'Inspector', depending on which is selected. 
* Can move fields within a location (like in the 'Editor' fields)
* Can move fields from one location to another (like from the 'Editor' fields to 'Inspector' fields)
* It was already possible to delete fields

#### Tickets
* [GF-2526](https://wpengine.atlassian.net/browse/GF-2526) (Add/remove fields)
* [GF-2528](https://wpengine.atlassian.net/browse/GF-2528) (Move/resize fields)
* [GF-2530](https://wpengine.atlassian.net/browse/GF-2530) (Inspector component)

#### Testing instructions
1. `composer install && npm install && npm run dev`
1. `/wp-admin` > Custom Blocks > Add New
1. Expected: something like this: <img width="1117" alt="bosf-sdf" src="https://user-images.githubusercontent.com/4063887/100181926-fa111280-2ea0-11eb-91ab-af14ff51a43c.png">
1. Please test deleting fields:
![the-filba](https://user-images.githubusercontent.com/4063887/100182299-e5814a00-2ea1-11eb-9ead-791ba90bee57.gif)
1. Create another new block, clicking `/wp-admin` > Custom Blocks > Add New
1. Please test moving fields. Moving isn't smooth like in Gutenberg, but it should work: ![creating-and-bo](https://user-images.githubusercontent.com/4063887/100785762-bce3dd80-33d6-11eb-9d10-7e89e7f702ba.gif)
1. Please test changing the location: 
![new-loc](https://user-images.githubusercontent.com/4063887/100786524-e6513900-33d7-11eb-9393-82ae1a1000af.gif)
1. Please test resizing: 
![chang-thidw](https://user-images.githubusercontent.com/4063887/100787684-a0957000-33d9-11eb-9d29-125f1efb21f5.gif)

kienstra committed:
Add the field to the right location

If it's in the inspector display,
add it to the inspector.

kienstra committed:
Change the signature of getNewFieldNumber to return a number of null

When it's used,
the return value calls .toString(),
so it can return a number.

kienstra committed:
Allow creating a new block, handle default slug (name)

On creating an saving (updating) a new post,
if there's no slug (name),
set a default one.

kienstra committed:
Add a titlePlaceholder to match the design

kienstra committed:
Merge in add/edit-block-ui-main, resolve conflict

There was only a conflict in
<KeywordsSection>

kienstra committed:
Break the useEffect hook into 2

It had a lot of dependencies
that other parts of the hook
did not need.

kienstra committed:
Move the setting of richEditingEnabled to PHP

No need to add in both JS and PHP,
just in PHP should be fine.

kienstra commented:
https://github.com/studiopress/genesis-custom-blocks/pull/47 should fix the failed PHPUnit test.

kienstra committed:
Create a helper function to get the default block

kienstra committed:
Prevent the editor state from being dirty on page load

When clicking /wp-admin > Custom Blocks >
Add New, the editor state shouldn't
allow an 'undo', as there
was nothing done.

kienstra committed:
Add the post ID to the defualt block name

If no name is entered,
this should use the name.

kienstra committed:
Add the moving buttons that Rob created

From Rob's markup.
Todo: make them move the field.

kienstra committed:
Allow moving fields, though the UI isn't ideal yet

This could probably use some kind of
UI to show that it moved,
maybe like in Gutenberg.

kienstra committed:
Don't allow moving a field if it's at the ends

If it's the first field,
disable the button that moves it up.

kienstra committed:
Will revert most: add tailwind as a CSS file

kienstra committed:
Make moving fields more intuitive

When the field is moved,
the up and down buttons appear
on the moved field.
It still could probably use
improvement.

kienstra committed:
Improve accesibility of <ClipboardCopy>

Add an aria-describedby attribute
and remove the custom styling
in favor of native tailwind.

kienstra committed:
Fix the warning on running Jest tests

kienstra commented:
This is mainly working:

![can-add-and-delete](https://user-images.githubusercontent.com/4063887/100630398-1f15e300-32f0-11eb-81ac-e9c8170c5d19.gif)

kienstra committed:
Add a helper function to get the other location

A little better than storing
the selected location as a boolean.
This should be easier to understand.

kienstra committed:
Extract a helper function to get the proper order

kienstra committed:
Merge branch 'feature/react-ui' into add/allow-add-remove-fields

kienstra committed:
Re-enable moving fields

The previous commit saved moved
fields as an array, but they
should be an object.

kienstra commented on php/Admin/EditBlock.php:
```diff
@@ -111,51 +111,32 @@ public function enqueue_assets() {
 		wp_enqueue_style(
 			self::STYLE_SLUG,
-			$this->plugin->get_url( 'css/edit-block.css' ),
+			$this->plugin->get_url( $edit_block_style_path ),
```
Todo: remove this, it's just for development.

kienstra commented on php/Admin/EditBlock.php:
```diff
@@ -111,51 +111,32 @@ public function enqueue_assets() {
-	 *
-	 * @return array|null The initial edits.
-	 */
-	public function get_initial_edits() {
```
This probably doesn't apply to us. 

If it's a new post, we have defaults ready to add, like a title of 'New Field' if they don't enter one.

kienstra committed:
Add

kienstra committed:
Move addNewField() to useField() hook

This probably makes more sense
than in the useBlock() hook.

kienstra committed:
Don't expose changeFieldName() to components

Ideally, this should
take care of the name change inside the hook.

kienstra committed:
Handle case of deleting entire field name

If someone deletes the slug,
the field should still display.

kienstra committed:
When adding a new field, display it in the <Side>

So the user knows
where to edit the field.

kienstra committed:
Revert autoslugging, out of scope for this PR

To be added in an upcoming PR.

kienstra committed:
Fix an error if there is no field

On creating a new block,
if you clicked the 'Field' panel,
there was an error.

kienstra committed:
Revert copying Tailwind into this repo

Eventually, we should have a build
system for tailwind,
and not hotlink it.

kienstra commented on js/src/common/helpers/getFieldsAsArray.js:
```diff
@@ -24,10 +24,6 @@ const compare = ( a, b ) => {
 	return Object.keys( fields )
 		.reduce( ( accumulator, fieldName ) => {
-			if ( '' === fieldName ) {
-				return accumulator;
```
A field might not have a name, like if a user deletes the name (slug) before re-entering it. 

They should still see the field, even though it has no name.

kienstra replied:
I might have to look into regressions this could cause in the normal block editor.

kienstra commented on js/src/edit-block/components/editor.js:
```diff
@@ -40,23 +43,37 @@ import { BrowserURL, Header, Main, Side } from './';
+	const { block, changeBlockName } = useBlock();
+	const [ selectedField, setSelectedField ] = useState( null );
+	const [ currentLocation, setCurrentLocation ] = useState( DEFAULT_LOCATION );
+	const [ panelDisplaying, setPanelDisplaying ] = useState( BLOCK_PANEL );
```
Eventually, maybe `useReducer()` instead of 3 `useState()` calls.

kienstra replied:
Hm, maybe not:

>When it's just an independent element of state you're managing: useState

https://kentcdodds.com/blog/should-i-usestate-or-usereducer

kienstra committed:
Reorder functions in useField(), no functional change intended

kienstra commented on js/src/block-editor/components/fields.js:
```diff
@@ -47,7 +47,6 @@ const getClassName = ( field ) => {
  * @property {Object} parentBlock The block where the fields are.
  * @property {Object} parentBlockProps The props to pass to the control function.
  * @property {number} [rowIndex] The index of the repeater row, if this field is in one (optional).
- * @property {Function} [changeFieldSetting] Edits a field value.
```
Not related to this PR, this prop just isn't needed.

kienstra commented on js/src/edit-block/helpers/convertToSlug.js:
```diff
@@ -0,0 +1,15 @@
+ * @return {string} The string as a slug, like 'example-here'.
+ */
+const convertToSlug = ( toConvert ) => {
+	return toConvert
```
Not used now, but will be useful in the upcoming 'autoslugging' tickets for fields and blocks.

kienstra commented on js/src/edit-block/hooks/useField.js:
```diff
@@ -54,58 +104,168 @@ const useField = () => {
+	const changeFieldSettings = useCallback( ( fieldName, newSettings ) => {
+		if ( newSettings.location ) {
+			fullBlock[ blockNameWithNameSpace ].fields = changeFieldLocation(
+				fullBlock[ blockNameWithNameSpace ].fields,
```
If a field's location is changed, it should be placed at the end of the order in the location.

For example, if it's changed from 'editor' to 'inspector', and there are 3 'inspector' fields, it should be the 4th 'inspector' field.

kienstra commented on js/src/edit-block/hooks/useField.js:
```diff
@@ -54,58 +104,168 @@ const useField = () => {
+		}
+
+		if ( newSettings.hasOwnProperty( 'name' ) ) {
+			fullBlock[ blockNameWithNameSpace ].fields = changeFieldName(
```
Changing the name of a field is complex because the name is the key in the `fields` object, as well as the name property:

<img width="627" alt="field-name" src="https://user-images.githubusercontent.com/4063887/100783637-a720e900-33d3-11eb-9c65-98973972a84f.png">

kienstra committed:
Change the param tag for selectedField to include null

kienstra committed:
Use the passed location instead of a hard-coded one

kienstra committed:
Don't display buttons if there's only one field

Both up and down will be disabled,
so no need to display the buttons.

kienstra committed:
Fix an issue in changing the type

Before, on changing the type,
it didn't appear anymore.

kienstra committed:
Remove wordpress/env, as e2e tests are reverted right now

kienstra committed:
Update wordpress/scripts, see if that helps handing npm run test:js

kienstra committed:
Use jest instead of test-unit-js

Attempt to fix that hanging:
https://app.circleci.com/pipelines/github/studiopress/genesis-custom-blocks/886/workflows/123bf966-57c3-4512-8382-9fb029224947/jobs/6509

kienstra committed:
See if npm run test:js runs when breaking up command

kienstra committed:
See if removing the nodegit-workaround helps Jest run

kienstra committed:
Revert "Use jest instead of test-unit-js"

This reverts commit 21fbf9286867c30eb4068834c0baa7453e555ba0.

kienstra committed:
Try the latest version of Node for Jest tests

kienstra committed:
Go back to lts version, put back nodegit-workaround

kienstra committed:
Will probably revert: add verbose output to Jest

kienstra committed:
Try setting maxWorkers, per post

https://vgpena.github.io/jest-circleci/

kienstra commented on .circleci/config.yml:
```diff
@@ -95,7 +95,7 @@ jobs:
           node-version: lts
       - nodegit-workaround
-      - run: HUSKY_SKIP_INSTALL=1 npm ci && npm run test:js
+      - run: HUSKY_SKIP_INSTALL=1 npm ci && npm run test:js -- --maxWorkers=2
```
As suggested by https://vgpena.github.io/jest-circleci/

kienstra commented:
This is ready for review, and the [testing instructions](https://github.com/studiopress/genesis-custom-blocks/pull/48#issue-526175515) are up to date.

kienstra commented on js/src/edit-block/components/clipboard-copy.js:
```diff
@@ -43,16 +43,28 @@ const ClipboardCopy = ( { text } ) => {
+					__( 'Copy the field name of %1$s', 'genesis-custom-blocks' ),
+					text
+				) }
+			</span>
```
Not much of a change here, other than indentation and adding this `<span>` for accessibility.

kienstra commented on js/src/edit-block/components/editor.js:
```diff
@@ -74,12 +91,7 @@ const Editor = ( { initialEdits, onError, postId, postType, settings } ) => {
-					settings={
-						{
-							...settings,
-							richEditingEnabled: false,
```
This is now set via PHP

kienstra commented on js/src/edit-block/components/fields-grid.js:
```diff
@@ -124,42 +82,109 @@ const FieldsGrid = ( { setSelectedFieldName } ) => {
-										</button>
-										<button className="field-resize w-4 absolute -mr-2 right-0 top-0 bottom-0 focus:outline-none"></button>
+										</div>
+										{ shouldDisplayMoveButtons
```
If there's only 1 field, there's no need to display move buttons.

kienstra commented on js/src/edit-block/hooks/useBlock.js:
```diff
@@ -33,21 +35,58 @@ const useBlock = () => {
+	const blockNameWithNameSpace = getBlockNameWithNameSpace( fullBlock );
+	const block = useMemo(
+		() => fullBlock[ blockNameWithNameSpace ] || {},
+		[ fullBlock, blockNameWithNameSpace ]
```
This way, it will only calculate `block` when `fullBlock` or `blockNameWithNameSpace` change.

marksabbath reviewed (changes requested):
When trying to delete a block using the instructions [here](https://github.com/studiopress/genesis-custom-blocks/pull/48#issuecomment-735864304) I get the error below @kienstra 

```
TypeError: Cannot read property 'settings' of undefined
    at FieldSettings (http://wptest.local/wp-content/plugins/genesis-custom-blocks/js/dist/edit-block.js?ver=46eae680c36421009a5097052e5a86c7:7068:158)
    at je (http://wptest.local/wp-includes/js/dist/vendor/react-dom.min.js?ver=16.9.0:78:476)
    at qe (http://wptest.local/wp-includes/js/dist/vendor/react-dom.min.js?ver=16.9.0:89:115)
    at ph (http://wptest.local/wp-includes/js/dist/vendor/react-dom.min.js?ver=16.9.0:216:489)
    at lh (http://wptest.local/wp-includes/js/dist/vendor/react-dom.min.js?ver=16.9.0:126:409)
    at O (http://wptest.local/wp-includes/js/dist/vendor/react-dom.min.js?ver=16.9.0:121:71)
    at ze (http://wptest.local/wp-includes/js/dist/vendor/react-dom.min.js?ver=16.9.0:118:14)
    at http://wptest.local/wp-includes/js/dist/vendor/react-dom.min.js?ver=16.9.0:53:49
    at unstable_runWithPriority (http://wptest.local/wp-includes/js/dist/vendor/react.min.js?ver=16.9.0:26:340)
    at Ma (http://wptest.local/wp-includes/js/dist/vendor/react-dom.min.js?ver=16.9.0:52:280)
```

It looks like the problem is happening _after_ the field is deleted and the frontend is trying to render an `undefined`.

By adding a `control &&` right before [this](https://github.com/studiopress/genesis-custom-blocks/pull/48/files#diff-16621d8366f81696fbe9daa9e68998d620a694466d9c695c1d7ad43f0a975ec0R48) resolves, however, the `undefined` is still being triggered. 

Tested using Local, WordPress 5.5.3, the twentytwenty theme, and no other plugin active, Chrome in the latest version which is  86.0.4240.198 (Official Build) (x86_64).

kienstra commented:
Hi @marksabbath,
Thanks for your great details, I'll work on this.

kienstra committed:
Handle a deleted field by not rendering settings

As Marcos mentioned, when deleting a field,
there was an error:
https://github.com/studiopress/genesis-custom-blocks/pull/48#pullrequestreview-542835794
Also, remove the selection of the field
so that it doesn't display an empty field.

kienstra committed:
Reorder constants import alphabetically

kienstra commented:
Hi @marksabbath,
Thanks for pointing that out, I also saw that error on deleting. Do the latest commits fix that in your local?

kienstra commented on js/src/edit-block/components/field-settings.js:
```diff
@@ -70,11 +73,15 @@ const FieldSettings = ( {
-					onClick={ deleteField }
+					onClick={ () => {
+						deleteField();
+						setSelectedField( NO_FIELD_SELECTED );
```
When a field is deleted, it should no longer be the `selectedField`.

marksabbath commented:
> Hi @marksabbath,
> Thanks for pointing that out, I also saw that error on deleting. Do the latest commits fix that in your local?

Yes! I don't see the issue anymore 🥳

kienstra commented:
Great, thanks for reviewing this!
